File: tasks/translation/output/Qwen/Qwen2.5-Coder-32B-Instruct-AWQ/codenet/direct/comment/uncomment/direct/Java/Go/s316260310_commented.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const (
	MAXN = 1000006
	MOD  = 1000000007
)

type Helper struct {
	mod   int64
	maxn  int
	rnd   *rand.Rand
	in    *bufio.Scanner
	out   *bufio.Writer
	buf   []byte
	index int
	total int
}

func NewHelper(mod int64, maxn int) *Helper {
	h := &Helper{
		mod:  mod,
		maxn: maxn,
		rnd:  rand.New(rand.NewSource(time.Now().UnixNano())),
	}
	h.buf = make([]byte, 1<<20)
	return h
}

func (h *Helper) initIO(in *os.File, out *os.File) {
	h.in = bufio.NewScanner(in)
	h.in.Buffer(h.buf, len(h.buf))
	h.out = bufio.NewWriter(out)
}

func (h *Helper) next() string {
	h.in.Scan()
	return h.in.Text()
}

func (h *Helper) nextInt() int {
	s := h.next()
	i, _ := strconv.Atoi(s)
	return i
}

func (h *Helper) nextLong() int64 {
	s := h.next()
	i, _ := strconv.ParseInt(s, 10, 64)
	return i
}

func (h *Helper) getIntArray(size int) []int {
	arr := make([]int, size)
	for i := 0; i < size; i++ {
		arr[i] = h.nextInt()
	}
	return arr
}

func (h *Helper) println(a ...interface{}) {
	fmt.Fprintln(h.out, a...)
}

func (h *Helper) flush() {
	h.out.Flush()
}

type Solver struct {
	hp         *Helper
	TESTCASES  bool
	timer      *time.Timer
	task       *time.Timer
	straight   []int
	cycle      []int
	vis        []bool
	curr       int
	st         int
	N          int
	K          int64
	A          []int
}

func NewSolver(hp *Helper) *Solver {
	s := &Solver{
		hp:        hp,
		TESTCASES: false,
	}
	return s
}

func (s *Solver) solve() {
	tc := 1
	if s.TESTCASES {
		tc = s.hp.nextInt()
	}
	for tce := 1; tce <= tc; tce++ {
		s.solveTc(tce)
	}
	s.timer.Stop()
	s.hp.flush()
}

func (s *Solver) solveTc(tc int) {
	s.N = s.hp.nextInt()
	s.K = s.hp.nextLong()
	s.A = s.hp.getIntArray(s.N)

	s.straight = []int{}
	s.cycle = []int{}
	s.vis = make([]bool, s.N)
	s.curr = 0
	for !s.vis[s.curr] {
		s.vis[s.curr] = true
		s.curr = s.A[s.curr] - 1
	}

	s.st = 0
	for s.st != s.curr {
		s.straight = append(s.straight, s.st)
		s.st = s.A[s.st] - 1
	}

	s.cycle = append(s.cycle, s.curr)
	s.curr = s.A[s.curr] - 1
	for s.curr != s.cycle[0] {
		s.cycle = append(s.cycle, s.curr)
		s.curr = s.A[s.curr] - 1
	}

	if s.K < int64(len(s.straight)) {
		s.hp.println(s.straight[s.K] + 1)
	} else {
		s.K -= int64(len(s.straight))
		s.hp.println(s.cycle[s.K%int64(len(s.cycle))] + 1)
	}
}

func main() {
	hp := NewHelper(MOD, MAXN)
	hp.initIO(os.Stdin, os.Stdout)
	solver := NewSolver(hp)
	solver.solve()
}

// End of Code

File: tasks/translation/output/Qwen/Qwen2.5-Coder-32B-Instruct-AWQ/codenet/direct/comment/uncomment/direct/Java/C++/s316260310_commented.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
const int MAXN = 1000006;
const ll MOD = 1e9 + 7;

class Helper {
public:
    Helper(ll mod, int maxn) : MOD(mod), MAXN(maxn) {}

    ll gcd(ll a, ll b) {
        return b == 0 ? a : gcd(b, a % b);
    }

    ll max(ll a, ll b) {
        return a > b ? a : b;
    }

    ll min(ll a, ll b) {
        return a < b ? a : b;
    }

    ll sum(vector<ll>& ar) {
        ll sum = 0;
        for (ll itr : ar) sum += itr;
        return sum;
    }

    void shuffle(vector<int>& ar) {
        random_shuffle(ar.begin(), ar.end());
    }

    void shuffle(vector<ll>& ar) {
        random_shuffle(ar.begin(), ar.end());
    }

    void reverse(vector<int>& ar) {
        reverse(ar.begin(), ar.end());
    }

    void reverse(vector<ll>& ar) {
        reverse(ar.begin(), ar.end());
    }

    ll pow(ll base, ll exp, ll MOD) {
        base %= MOD;
        ll ret = 1;
        while (exp > 0) {
            if (exp & 1) ret = ret * base % MOD;
            base = base * base % MOD;
            exp >>= 1;
        }
        return ret;
    }

    void initIO(istream& is, ostream& os) {
        cin.rdbuf(is.rdbuf());
        cout.rdbuf(os.rdbuf());
    }

    string next() {
        string s;
        cin >> s;
        return s;
    }

    int nextInt() {
        int x;
        cin >> x;
        return x;
    }

    ll nextLong() {
        ll x;
        cin >> x;
        return x;
    }

    void print(const string& a) {
        cout << a;
    }

    void printsp(const string& a) {
        print(a);
        print(" ");
    }

    void println() {
        cout << "\n";
    }

    void println(const string& a) {
        print(a);
        println();
    }

    void flush() {
        cout.flush();
    }

private:
    ll MOD;
    int MAXN;
};

class Solver {
public:
    Solver() {
        hp = Helper(MOD, MAXN);
        hp.initIO(cin, cout);
    }

    void solve() {
        int tc = TESTCASES ? hp.nextInt() : 1;
        for (int tce = 1; tce <= tc; ++tce) solve(tce);
    }

    bool TESTCASES = false;

    void solve(int tc) {
        int N = hp.nextInt();
        ll K = hp.nextLong();
        vector<int> A(N);
        for (int i = 0; i < N; ++i) A[i] = hp.nextInt();

        vector<int> straight, cycle;
        vector<bool> vis(N, false);
        int curr = 0;
        while (!vis[curr]) {
            vis[curr] = true;
            curr = A[curr] - 1;
        }

        int st = 0;
        while (st != curr) {
            straight.push_back(st);
            st = A[st] - 1;
        }

        cycle.push_back(curr);
        curr = A[curr] - 1;
        while (curr != cycle[0]) {
            cycle.push_back(curr);
            curr = A[curr] - 1;
        }

        if (K < straight.size()) {
            hp.println(to_string(straight[K] + 1));
        } else {
            K -= straight.size();
            hp.println(to_string(cycle[K % cycle.size()] + 1));
        }
    }

private:
    Helper hp;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    Solver solver;
    solver.solve();
    return 0;
}
// End of Code
